### ISSUE TITLE
Pure functions

### ISSUE THREAD

**Issue by CGamesPlay** [facebook/flow#2715]
This is a feature request for marking functions as not having any side effects, aka being pure.

Flow reports the following safe code as unsafe:

```
/* @flow */
type T = { func: ?() => void };
function f(): void { }
function g(val: T) {
  if (val.func) {
    f();
    val.func();
  }
}
```

The reasoning for this is because Flow has no idea who else has a reference to `val`, and any function call might modify the value of `val.func`, so the previous non-null assertion is invalidated.

It would be nice to be able to mark functions as `pure`. Flow could assert that:
- no variables outside the local scope are modified (including properties of objects)
- no non-pure functions are called

As a straw man syntax for this, I suggest `pure` before the return type in the type definition:

```
type APureFunction: () => pure void
function aPureFunction(): pure void { }
```

I don't think this is a duplicate of #18 because that focuses on data immutability, which is different from function purity.

**Comment by CGamesPlay** [comment]
As a potentially more realistic example of why the current behavior is annoying, this is closer to what I was trying to do when I stumbled upon this limitation.

```
/* @flow */
type T = { func: ?(number) => void };
function f() { return 7; }
function g(val: T) {
  if (val.func) {
    val.func(f());
  }
}
```

**Comment by bradennapier** [comment]
Just ran into this as well.  Especially when it comes to using `%checks` value as it means that every time you use any `%checks` function you have to redo every single refine you have already done.

**Comment by jcready** [comment]
@bradennapier it seems like predicate functions (using `%checks`) only cause previous refinements to be invalidated if they are called in a different statement from the original refinement ([Try Flow](https://flow.org/try/#0MYewdgzgLgBAlhAggJ2TAvDAFADwFwwC2cOApgCYCUBARiCADakCGYMApMABanADWEDAD4YKZMwCeAOgRjJuSgChQkWAgDyNAFYZs+IiQrUYdRizace-QehFQJAB1IgAZjBwZ0mAOQhtvKG8YADJg9xgAQi8YMABXBgYQsIjZVAVlcGh4CABlKDRMXAJiMipaeiZWDm5eAWEYeydXcK8faGQ4MABzbwzVbIA5WMJdIoNS41NKixrresbnNw9WmG84whpSZF6VLIQAIQrR-RKjcrMqy1qbO0dFlujvKfNevqyoUizC-QBvGABtdqdLoAXWKhnIMAAvsYgd16j9FDB4G4sAtmjgpC56J42vlgd5KDBEcjkXBUeillIaMwCo84T0iSTScjkKQoLFkGxMdiQDAAPT8mAOZgQCCfJGkqGS6XIyWCmAABTZ5DgwGYHxgLliYGAUDgmRg6ti4pgbJcnVIhFIYDUYAAbswGHByBqDZAUfA2KqXC4tjbYNANVaAxBJeTsJT3FicStvAzCcTJWTUQg8shcNTaZQmcnSWyOVzo7yBUKXMw4AwwyzZdDJeHU7l8pneTmkyyI1g083MTTkG3mSyC5zuTG+Qry5Xq1KZfWU9huxmefQB3nO1He7TcasE7mWaz2SPiziFSKxRKa7O5ciFYcoFwzaQLWAQ7bBJ0GjwYBBmNbv1Bg2tW0YHIPkwBAWBjVNc1LSAu1HWdV19XABtIzuDEx23eN8W6IJQmydNMz7Vch0PItl3HIUz3Fada1QrsmyXMciXwxciOzPd8zI0cS1PUUaKvGBJWHItvF6KEgA))

```js
const isArr = (x: mixed): boolean %checks => Array.isArray(x)
const isObj = (x: mixed): boolean %checks => typeof x === 'object' && x !== null && !isArr(x)
const isStr = (x: mixed): boolean %checks => typeof x === 'string'
const isNum = (x: mixed): boolean %checks => typeof x === 'number'
const isBool = (x: mixed): boolean %checks => typeof x === 'boolean'

const test = (x: { [string]: mixed }): string => {
  if (typeof x.foo === 'string') {
    if (typeof x.bar === 'string') {
      return x.foo // passes
    }
  }
  
  // Predicate functions cause refinement invalidations if in different statements
  if (typeof x.foo === 'string') {
    if (isStr(x.bar)) {
      return x.foo // fails
    }
  }
  
  if (isStr(x.foo)) {
    if (isStr(x.bar)) {
      return x.foo // fails
    }
  }
  
  if (isStr(x.foo)) {
    if (typeof x.bar === 'string') {
      return x.foo // passes
    }
  }
  
  // Both refinements in the same statement do not cause refinement invalidation
  if (typeof x.foo === 'string' && isStr(x.bar)) {
    return x.foo // passes
  }
  
  if (isStr(x.foo) && isStr(x.bar)) {
    return x.foo // passes
  }
  
  return ''
}
```

**Comment by bradennapier** [comment]
Yeah, which is the problem.  "Pure" functions as a concept would relieve this as you are essentially promising Fllow that no mutations or side effects will occur when a given function is called, so it should not invalidate refinements when called.  

I would think %checks would default to being "pure" since its a simple boolean operation whereas there would be a way to mark standard functions as pure (perhaps jsut with `%pure` in place of `%checks`).

**Comment by jcready** [comment]
But do you not find it odd that the refinements are **not** invalidated if you just put the function call in the same statement?

```js
if (typeof x.foo === 'string' && isStr(x.bar)) {
  return x.foo // passes
}
```

**Comment by bradennapier** [comment]
Yeah, I think that this may explain a few of the bugs going on right now with refinement actually.

**Comment by Squeegy** [comment]
#5441 also

**Comment by kevinbarabash** [comment]
This issue seems to be focused on side-effects that mutate data somewhere.  Pure functions in general aren't supposed to have any side-effects, e.g. throwing exceptions, logging to the console, drawing to the canvas, etc.  I think we should only mark functions as mutating or not.

**Comment by rattrayalex** [comment]
Agreed, being able to configure whether you care about things like logging would be ideal!
